File: coding-test/programmers/src/kr/co/programmers/greedy/GymSuit.java
package kr.co.programmers.greedy;

public class GymSuit {

    public static void main(String[] args) {
        int n = 5;
        int[] lost = {2, 4};
        int[] reserve = {1, 3, 5};
        System.out.println(new Solution().solution(n, lost, reserve));
    }

    static class Solution {
        public int solution(int n, int[] lost, int[] reserve) {

            int answer = 0;
            int[] students = new int[n];

            for(int l : lost) students[l-1]--;

            for(int r : reserve) students[r-1]++;

            for(int i = 0; i < students.length; i++){
                if(students[i] == -1){
                    if(i > 0 && students[i-1] > 1){
                        students[i-1]--;
                        students[i]++;
                        answer++;
                        continue;
                    }

                    if(i < students.length - 1 && students[i+1] > 1){
                        students[i+1]--;
                        students[i]++;
                        answer++;
                        continue;
                    }
                }
                answer++;
            }

            return answer;
        }
    }
}

File: coding-test/nodejs/src/algorithm/greedy/gym-suit.js
// @ts-check

/**
 *
 * @param {number} n
 * @param {number[]} lost
 * @param {number[]} reserve
 * @returns {number}
 */
function solution(n, lost, reserve) {
  let answer = 0

  const students = new Array(n)
  students.fill(1)

  for (let i = 0; i < lost.length; i += 1) {
    students[lost[i] - 1] -= 1
  }

  for (let i = 0; i < reserve.length; i += 1) {
    students[reserve[i] - 1] += 1
  }

  for (let i = 0; i < students.length; i += 1) {
    if (students[i] < 1) {
      if (i > 0 && students[i - 1] > 1) {
        students[i - 1] -= 1
        students[i] += 1
        answer += 1
        continue
      }

      if (i < students.length - 1 && students[i + 1] > 1) {
        students[i + 1] -= 1
        students[i] += 1
        answer += 1
        continue
      }
    } else {
      answer += 1
    }
  }

  return answer
}

const n = 5
const lost = [2, 4]
const reserve = [3]

console.log(solution(n, lost, reserve))
